### DISCUSSION TITLE
IModelHost.startup doesn't initialize hubAccess and causes downloadBriefcase to crash

### DISCUSSION THREAD

**Discussion by benjamin-lindsey** [iTwin/itwinjs-core#3424] [Q&A] [answered]
Hi,

I'm trying to download a briefcase using the `BriefcaseManager.downloadBriefcase` function from a backend service and get the following error: 
`<rejected> TypeError: Cannot read properties of undefined (reading 'getChangesetFromVersion')`

The reason? `IModelHost.hubAccess` is `undefined`. Yes, I've called `IModelHost.startup()` beforehand.

Looking through iTwin.js source code I find these two lines in `IModelHost.startup()` : 
https://github.com/iTwin/itwinjs-core/blob/aeb8ae8c894b51dd31d518da5a1ca25258d59f67/core/backend/src/IModelHost.ts#L411-L412

Those are the only two lines in the startup function (that I'm able to find) where `hubAccess` is referenced. If the `hubAccess` parameter is `undefined`, then it appears to never be set.

If we jump to `BriefcaseManager.downloadBriefcase` at this line: https://github.com/iTwin/itwinjs-core/blob/aeb8ae8c894b51dd31d518da5a1ca25258d59f67/core/backend/src/BriefcaseManager.ts#L210
Then we see that `hubAccess` is being used, even though it's `undefined`. There doesn't seem to be a point where `hubAccess` is checked for `undefined` before calling `getChangesetFromVersion`.

Is this an iTwin.js bug or am I doing something wrong?

Thanks!

**Comment by mgooding** [comment, marked as answer]
If you are going to access iModelHub from iTwin.js, you need to specify hubAccess in your IModelHostConfiguration. Something like:

```ts
import { BackendIModelsAccess } from "@itwin/imodels-access-backend";

const imhConfig: IModelHostConfiguration = {
  hubAccess: new BackendIModelsAccess(), // needed to download iModels from iModelHub
  // Required but unrelated fields...
  logTileLoadTimeThreshold: IModelHostConfiguration.defaultLogTileLoadTimeThreshold,
  logTileSizeThreshold: IModelHostConfiguration.defaultLogTileSizeThreshold,
  tileContentRequestTimeout: IModelHostConfiguration.defaultTileRequestTimeout,
  tileTreeRequestTimeout: IModelHostConfiguration.defaultTileRequestTimeout,
};
await IModelHost.startup(imhConfig);
```

I agree that this is confusing - I hit the same issue when upgrading one of my sample applications to iTwin.js 3.0. It's a side effect of untangling the iModelHub client from the core packages, but I think we will look to improve the ergonomics here in the future.

When I hit this issue, I submitted a PR that will at least provide a better error message for future users: https://github.com/iTwin/itwinjs-core/pull/3391

**Reply by benjamin-lindsey** [reply]
Woohoo, that was it! Thanks!

After running into an authentication error, it took a bit to figure out that [new client scopes](https://github.com/iTwin/viewer/blob/master/CHANGELOG-2.0.md#new-scopes-required) for 3.0 had to be added and old ones removed. Now it's smooth sailing.
